pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -1060,6 +1031,72 @@ func applyUpstreamTLSSettings(opts *buildClusterOpts, tls *networking.ClientTLSS
 				tlsContext.CommonTlsContext.AlpnProtocols = util.ALPNInMesh
 			}
 		}
+
+	case networking.ClientTLSSettings_SIMPLE, networking.ClientTLSSettings_MUTUAL:
+		if tls.Mode == networking.ClientTLSSettings_SIMPLE {
+			tlsContext = &auth.UpstreamTlsContext{
+				CommonTlsContext: &auth.CommonTlsContext{
+					ValidationContextType: &auth.CommonTlsContext_ValidationContext{
+						ValidationContext: certValidationContext,
+					},
+				},
+				Sni: tls.Sni,
+			}
+		} else if tls.Mode == networking.ClientTLSSettings_MUTUAL {
+			if tls.ClientCertificate == "" || tls.PrivateKey == "" {
+				log.Errorf("failed to apply tls setting for %s: client certificate and private key must not be empty",
+					c.Name)
+				return
+			}
+			tlsContext = &auth.UpstreamTlsContext{
+				CommonTlsContext: &auth.CommonTlsContext{},
+				Sni:              tls.Sni,
+			}
+
+			// Fallback to file mount secret instead of SDS if meshConfig.sdsUdsPath isn't set or SDS is disabled
+			if !node.Metadata.SdsEnabled || opts.push.Mesh.SdsUdsPath == "" {
```

Comment:
Not sure what's the point of this check- SDS is enabled by default and SdsUdsPath is deprecated, is this required for backward compatibility?
Also if we're serving files over SDS by default it should be documented somewhere for users as currently as a user only FileMount approach is explicitly mentioned. Was there a design doc for defaulting to file based SDS? With introduction of secret SDS shouldn't we fall back to simple file mount and not file based SDS?

Reply:
file based SDS is an implementation detail, the use does not need to know about it. Its strictly better than "simple file mount" in terms of the user experience (assuming it works correctly)

Reply:
cc: @JimmyCYJ

Reply:
> Not sure what's the point of this check- SDS is enabled by default and SdsUdsPath is deprecated, is this required for backward compatibility?

Nice catch! Perhaps this is outdated and is not executed nowadays. Could you add a TODO to deprecate this if condition? cc @myidpt to confirm.

> Also if we're serving files over SDS by default it should be documented somewhere for users as currently as a user only FileMount approach is explicitly mentioned. Was there a design doc for defaulting to file based SDS? With introduction of secret SDS shouldn't we fall back to simple file mount and not file based SDS?

As @howardjohn pointed out this filebased SDS is implementation details. But I don't know if there is a tracking GitHub issue for the filebased-SDS work, and I think this implementation does not have sufficient test coverage or performance analysis. If this filebased SDS is already completed and is in production quality, we need to update the user guide in istio.io. If there is any doc on file mount secret rotation, we need to update them, and remove the legacy filemount approach from code. cc @myidpt @lei-tang

Reply:
Since we have SDS enabled by default, the if block is outdated and not execute. We should clean it later. I added a TODO here for server certs https://github.com/istio/istio/blob/master/pilot/pkg/security/model/authentication.go#L177 and forgot here. Please leave a TODO we can clean them together.

Reply:
I added a TODO for the cleanup